File: src/lib/inci.ts
export interface ParsedIngredient {
  raw: string;
  normalized: string;
  position: number;
}

const STRIP_PARENS = /\s*\([^)]*\)\s*/g;

export function normalizeInci(name: string): string {
  return name
    .toLowerCase()
    .replace(STRIP_PARENS, ' ')
    .replace(/[*†‡]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

export function parseInci(raw: string): ParsedIngredient[] {
  if (!raw) return [];
  const seen = new Set<string>();
  const out: ParsedIngredient[] = [];
  const parts = raw
    .split(/[,;\n\r]+/)
    .map((s) => s.trim())
    .filter(Boolean);

  for (const original of parts) {
    const cleaned = original.replace(/^[.\-•·\d]+\s*/, '').trim();
    if (!cleaned) continue;
    const normalized = normalizeInci(cleaned);
    if (!normalized || seen.has(normalized)) continue;
    seen.add(normalized);
    out.push({ raw: cleaned, normalized, position: out.length });
  }
  return out;
}

File: ios/Skintel/SkintelCore/Sources/SkintelCore/Logic/INCI.swift
import Foundation

/// Exact port of src/lib/inci.ts. `inci_normalized` rows written by the web app were
/// produced by this normaliser, so the iOS app must produce byte-identical keys or
/// culprit correlation across web- and iOS-created products silently breaks.
public enum INCI {
    public struct ParsedIngredient: Sendable, Hashable {
        public var raw: String
        public var normalized: String
        public var position: Int

        public init(raw: String, normalized: String, position: Int) {
            self.raw = raw; self.normalized = normalized; self.position = position
        }
    }

    /// lowercase → strip "( … )" groups (replaced by a space) → strip * † ‡ → collapse whitespace → trim
    public static func normalize(_ name: String) -> String {
        var s = name.lowercased()
        s = s.replacing(/\s*\([^)]*\)\s*/, with: " ")
        s = s.replacing(/[*†‡]/, with: "")
        s = s.replacing(/\s+/, with: " ")
        return s.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Split on , ; and newlines, strip leading list markers/digits, dedupe on the normalised key.
    public static func parse(_ raw: String) -> [ParsedIngredient] {
        guard !raw.isEmpty else { return [] }
        var seen = Set<String>()
        var out: [ParsedIngredient] = []
        let parts = raw.split(separator: /[,;\n\r]+/)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        for original in parts {
            let cleaned = original.replacing(/^[.\-•·\d]+\s*/, with: "")
                .trimmingCharacters(in: .whitespacesAndNewlines)
            if cleaned.isEmpty { continue }
            let normalized = normalize(cleaned)
            if normalized.isEmpty || seen.contains(normalized) { continue }
            seen.insert(normalized)
            out.append(ParsedIngredient(raw: cleaned, normalized: normalized, position: out.count))
        }
        return out
    }
}
